### ISSUE TITLE
Get first related node query gives Neo.TransientError.General.StackOverFlowError

### ISSUE THREAD

**Issue by leng-lim** [neo4j/neo4j#12558]
## Environment

- Neo4j version: 4.0.0
- Operating system: Linux Mint 19.3 Cinnamon
- API/Driver: neo4j-graphql-java

## Steps to Reproduce

- Create the Movie Graph via the neo4j browser
- Run the following cypher query:

`MATCH (person:Person) RETURN person {actedIn:[(person)-[:ACTED_IN]->(movie:Movie) | movie { .title }][0]} AS person ORDER BY person.name ASC LIMIT 10`

## Expected Behaviour
A list of Person nodes sorted by name containing a field with with the first movie that they have acted in.

## Actual Behaviour
The following error is thrown:

**Neo.TransientError.General.StackOverFlowError**
`There is not enough stack size to perform the current task. This is generally considered to be a database error, so please contact Neo4j support. You could try increasing the stack size: for example to set the stack size to 2M, add `dbms.jvm.additional=-Xss2M' to in the neo4j configuration (normally in 'conf/neo4j.conf' or, if you are using Neo4j Desktop, found through the user interface) or if you are running an embedded installation just add -Xss2M as command line flag.`

## Additional Notes
Increasing the stack size by adding `dbms.jvm.additional=-Xss2M` makes it take longer to fail but I still get the same error. I even tried setting it to 100M. The query can be fixed by removing the "ORDER BY" clause or by not trying to only extract the first movie of the "ACTED_IN" relationship i.e. the following queries work:

- **Remove "ORDER BY" clause** - `MATCH (person:Person) RETURN person {actedIn:[(person)-[:ACTED_IN]->(movie:Movie) | movie { .title }][0]} AS person LIMIT 10`
- **Do not extract fist movie** - `MATCH (person:Person) RETURN person {actedIn:[(person)-[:ACTED_IN]->(movie:Movie) | movie { .title }]} AS person ORDER BY person.name ASC LIMIT 10`

For added context, we're using the Java neo4j-graphql-java library to translate GraphQL queries to Cypher so we don't have much control over the generated Cypher.

**Comment by pontusmelke** [comment]
Hi @leng-lim , thanks for reporting this issue.

We have been able to reproduce the issue and fix the issue. So if you upgrade the database (to at least version 4.0.4) it should be working fine again.

I'll close this issue but please feel free to reopen or open a new one if the problem reappear.

Regards
Pontus

**Comment by joelduerksen** [comment]
Is it possible to get a little information on how/why this issue arose?  and/or the fix?  
I'm seeing this error message randomly on neo4j 4.1.1, the same query will throw the error at times, and not others.
my db is a research non-production db with no other query traffic/load, and is static content

I know this may not help but it feels like there is a resource collector that gets behind.  I wait a while and the error goes away, for a bit, but is constant when it starts occuring.  I loaded the same db on two computers, one has 16 threads, the other 48 threads (this is where I first saw the error)
